File: evals/retrievers/keyword.ts
import type { Chunk } from '../lib/corpus.ts'
import { tokenize } from '../lib/tokenize.ts'

export type Retriever = { name: string; search(query: string, k: number): string[] }

/**
 * Mirrors the agent's keyword fallback (case-insensitive substring), which is the baseline
 * embeddings have to beat to justify their cost. Kept here rather than imported: src modules
 * use extensionless imports Node cannot resolve.
 */
export function keywordRetriever(chunks: Chunk[]): Retriever {
  const folded = chunks.map((c) => ({ id: c.id, hay: tokenize(c.text).join(' ') }))
  return {
    name: 'keyword',
    search(query, k) {
      const terms = tokenize(query)
      if (terms.length === 0) {
        return []
      }
      return folded
        .map((c) => ({ id: c.id, hits: terms.filter((t) => c.hay.includes(t)).length }))
        .filter((c) => c.hits > 0)
        .sort((a, b) => b.hits - a.hits)
        .slice(0, k)
        .map((c) => c.id)
    },
  }
}

File: evals/retrievers/keyword.py
from typing import Protocol

from lib.corpus import Chunk
from lib.tokenize import tokenize


class Retriever(Protocol):
    @property
    def name(self) -> str: ...

    def search(self, query: str, k: int) -> list[str]: ...


class KeywordRetriever:
    name = 'keyword'

    def __init__(self, chunks: list[Chunk]) -> None:
        self.folded = [(c.id, ' '.join(tokenize(c.text))) for c in chunks]

    def search(self, query: str, k: int) -> list[str]:
        terms = tokenize(query)
        if not terms:
            return []
        scored = [(id, sum(1 for t in terms if t in hay)) for id, hay in self.folded]
        hits = [(id, n) for id, n in scored if n > 0]
        hits.sort(key=lambda pair: pair[1], reverse=True)
        return [id for id, _ in hits[:k]]
